### PR TITLE
Req id review

### PR DESCRIPTION
**Pull request by kelvin-cao**
In current NTB driver, we add requester IDs for RC and USP dynamically during driver initialization. This overrides the requester IDs pre-configured from config file. 

This PR adds NTB driver the capability of awareness of pre-configured requester IDs and avoid overriding them when adding new ones. And, a sysfs file ntb/requester_ids is introduced for adding, deleting and showing requester IDs dynamically at run time. 

Besides the feature changes, a minor bug in config_req_id_table()  is also fixed.

### CODE REVIEW

File: ntb_hw_switchtec.c [hunk 1]
```diff
@@ -1041,6 +1041,61 @@ static int config_req_id_table(struct switchtec_ntb *sndev,
 	return 0;
 }
 
+static int add_req_ids(struct switchtec_ntb *sndev,
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1053]: This seems unnecessarily complicated just to add requester ids... We shouldn't need to allocate a chunk of memory to read back the entire list just to add to that list just to maintain the existing config_req_id_table() interface. And based on the current usage of config_req_id_table() we probably don't a method to add batch ids.

So, roughly, just have three functions:

* add_req_id() - which config_req_id_table() can be change to which just finds the first un-enabled entry on hardware and adds it

* del_req_id() which searches for an entry and disables it

* clr_req_ids() (if necessary) to remove all entries in the table during initialization.

File: ntb_hw_switchtec.c [hunk 2]
```diff
@@ -1573,6 +1617,146 @@ static int switchtec_ntb_reinit_peer(struct switchtec_ntb *sndev)
...
+				   const char *page, size_t count)
+{
+	int bus, dev, func;
+	int req_id;
+	bool add_req_id;
+	ssize_t ret;
+
+	if (sscanf(page, "add %x:%x.%x", &bus, &dev, &func) == 3) {
+		add_req_id = true;
+	} else if (sscanf(page, "del %x:%x.%x", &bus, &dev, &func) == 3) {
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1669]: I think this goes a bit against convention. There should be sysfs attributes for adding and removing. Similar to the bind and unbind attributes for attaching drivers. Better than having one attribute that does everything.

File: ntb_hw_switchtec.c [hunk 3]
```diff
@@ -1573,6 +1617,146 @@ static int switchtec_ntb_reinit_peer(struct switchtec_ntb *sndev)
...
+	.default_attrs = switchtec_ntb_attrs,
+};
+
+static void switchtec_ntb_init_sysfs(struct switchtec_ntb *sndev)
+{
+	struct kobject *parent;
+	int err;
+
+	parent = &sndev->stdev->dev.kobj;
+	err = kobject_init_and_add(&sndev->kobj, &switchtec_ntb_ktype, parent,
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1750]: This is very much the wrong API level. You should not be creating new kobjects. 

sysfs_create_group() would be more appropriate.

File: ntb_hw_switchtec.c [hunk 4]
```diff
@@ -125,6 +125,8 @@ struct switchtec_ntb {
 	enum ntb_width link_width;
 	struct work_struct check_link_status_work;
 	bool link_force_down;
+
+	struct kobject kobj;
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:129]: This violates the documentation:
>   No structure should **EVER** have more than one kobject embedded within it.
>  If it does, the reference counting for the object is sure to be messed
>  up and incorrect, and your code will be buggy.  So do not do this. [[kobject]]

switchtec_ntb has a kobj already in ntb.dev.kobj.

[kobject]: https://elixir.bootlin.com/linux/latest/source/Documentation/core-api/kobject.rst

File: ntb_hw_switchtec.c [hunk 5]
```diff
@@ -1573,6 +1617,146 @@ static int switchtec_ntb_reinit_peer(struct switchtec_ntb *sndev)
...
+	.sysfs_ops = &switchtec_ntb_sysfs_ops,
+	.default_attrs = switchtec_ntb_attrs,
+};
+
+static void switchtec_ntb_init_sysfs(struct switchtec_ntb *sndev)
+{
+	struct kobject *parent;
+	int err;
+
+	parent = &sndev->stdev->dev.kobj;
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1749]: I'd probably add these attributes to the NTB device (sndev->ntb->dev) and not the PCI device.

File: ntb_hw_switchtec.c (hunk 1 above)
Comment (kelvin-cao) [reply ntb_hw_switchtec.c:1053]: The reason I wanted to keep the config_req_id_table() as is is that I wanted to make the lock/unlock of NT operations just in this function, so that other functions like add, del can just focus on the requester ids handling without duplicating those NT operation code. With current code, the config_req_id_table() just program the table while other functions prepare the table. 

For adding a requester id, to prevent from adding duplicated one, it's necessary to go through the entire table and compare each enabled entry with the one to be added. So I think if we can sort the list of enabled entries when we go through them will benefit delete operations a bit, as it will only need to look at the first enabled ones.

And one benefit of addding batch ids with current code is that we can tell if this batch adding will success before we try insert each entry and finally find that there's not enough room for the remaining entries when we have inserted some.

File: ntb_hw_switchtec.c (hunk 2 above)
Comment (kelvin-cao) [reply ntb_hw_switchtec.c:1669]: I'll split it into three attributes 'add_requester_id', 'del_requester_id', and 'requester_ids' for adding, removing and showing.

File: ntb_hw_switchtec.c (hunk 3 above)
Comment (kelvin-cao) [reply ntb_hw_switchtec.c:1750]: Will update.

File: ntb_hw_switchtec.c (hunk 4 above)
Comment (kelvin-cao) [reply ntb_hw_switchtec.c:129]: Will update. Thanks!

File: ntb_hw_switchtec.c (hunk 5 above)
Comment (kelvin-cao) [reply ntb_hw_switchtec.c:1749]: Agree, will update.

File: ntb_hw_switchtec.c (hunk 1 above)
Comment (lsgunth) [reply ntb_hw_switchtec.c:1053]: I don't think we should worry too much about being efficient about lock and unlock operations as these are all on the slow path. It's certainly not worth the extra mess.

If you do want to combine the lock around the two cases that do a small amount of batch adding, then add a lock and unlock function that must be called around add_req_id and del_req_id().

File: ntb_hw_switchtec.c [hunk 6]
```diff
@@ -992,20 +992,84 @@ static int config_rsvd_lut_win(struct switchtec_ntb *sndev,
...
+	for (i = 0; i < table_size; i++) {
+		proxy_id = ioread32(&mmio_ctrl->req_id_table[i]);
 
-	if (ioread16(&mmio_ctrl->req_id_table_size) < count) {
+		if (!(proxy_id & NTB_CTRL_REQ_ID_EN) && (slot == -1))
+			slot = i;
+
+		if ((proxy_id & NTB_CTRL_REQ_ID_EN) &&
+		    (proxy_id >> 16) == req_id) {
+			exist = true;
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1024]: Maybe just goto an unlock tag, instead of adding the extra variable, condition and indentation.

File: ntb_hw_switchtec.c [hunk 7]
```diff
@@ -992,20 +992,84 @@ static int config_rsvd_lut_win(struct switchtec_ntb *sndev,
...
+	if (rc)
+		return rc;
+
+	iowrite32(NTB_PART_CTRL_ID_PROT_DIS, &mmio_ctrl->partition_ctrl);
+
+	for (i = 0; i < table_size; i++) {
+		proxy_id = ioread32(&mmio_ctrl->req_id_table[i]);
 
-	if (ioread16(&mmio_ctrl->req_id_table_size) < count) {
+		if (!(proxy_id & NTB_CTRL_REQ_ID_EN) && (slot == -1))
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1019]: All these extra braces are unecessary... (and below). You don't need braces around if condition in an if statement.

Just the first set is necessary for the "!".

File: ntb_hw_switchtec.c [hunk 8]
```diff
@@ -992,20 +992,84 @@ static int config_rsvd_lut_win(struct switchtec_ntb *sndev,
...
+			error);
 	}
 
+	if (!added)
+		return -EFAULT;
+
+	return 0;
+}
+
+static int cls_req_ids(struct switchtec_ntb *sndev,
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1121]: I don't think cls is a common abbreviation.... I assume this is clear? Make it 'clear' or 'clr'.

File: ntb_hw_switchtec.c [hunk 9]
```diff
@@ -1110,20 +1165,26 @@ static int crosslink_setup_mws(struct switchtec_ntb *sndev, int ntb_lut_idx,
...
+	for (i = 0; i < table_size; i++) {
 		proxy_id = ioread32(&sndev->mmio_self_ctrl->req_id_table[i]);
 
 		if (!(proxy_id & NTB_CTRL_REQ_ID_EN))
-			break;
+			continue;
 
-		req_ids[i] = ((proxy_id >> 1) & 0xFF);
+		proxy_id = ((proxy_id >> 1) & 0xFF);
+		rc = add_req_id(sndev, mmio_ctrl, proxy_id);
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1241]: Should probably handle the error here instead of ignoring rc.

File: ntb_hw_switchtec.c [hunk 10]
```diff
@@ -1324,20 +1385,22 @@ static void switchtec_ntb_init_msgs(struct switchtec_ntb *sndev)
...
 {
-	int req_ids[2];
+	int req_id;
+	int rc;
 
 	/*
 	 * Root Complex Requester ID (which is 0:00.0)
 	 */
-	req_ids[0] = 0;
+	req_id = 0;
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1451]: No need to assign to the temporary variable... just use 0 directly.

File: ntb_hw_switchtec.c [hunk 11]
```diff
@@ -1061,6 +1061,63 @@ static int add_req_id(struct switchtec_ntb *sndev,
 	return 0;
 }
 
+static int del_req_id(struct switchtec_ntb *sndev,
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1064]: I'd put this in another patch.

File: ntb_hw_switchtec.c [hunk 12]
```diff
@@ -1061,6 +1061,63 @@ static int add_req_id(struct switchtec_ntb *sndev,
...
+		rid >>= 16;
+		if (rid == req_id) {
+			iowrite32(0, &mmio_ctrl->req_id_table[i]);
+			break;
+		}
+	}
+
+	if (i == table_size) {
+		dev_err(&sndev->stdev->dev,
+			"Failed to delete requester ID %02X:%02X.%X.\n",
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1098]: Probably better to say note that the requester ID was not in the table. This suggests it is in the table and could not be removed.

File: ntb_hw_switchtec.c [hunk 13]
```diff
@@ -1581,6 +1638,104 @@ static int switchtec_ntb_reinit_peer(struct switchtec_ntb *sndev)
...
+	table_size = ioread16(&sndev->mmio_self_ctrl->req_id_table_size);
+
+	for (i = 0; i < table_size; i++) {
+		req_id = ioread32(&sndev->mmio_self_ctrl->req_id_table[i]);
+
+		if (!(req_id & NTB_CTRL_REQ_ID_EN)) {
+			n += sprintf(req_id_str, "Requester ID #%d: N/A\n", i);
+		} else {
+			req_id >>= 16;
+			n += sprintf(req_id_str, "Requester ID #%d: %02x:%02x.%x\n", i,
```
Comment (lsgunth) [inline ntb_hw_switchtec.c:1717]: Generally we don't want our sysfs files to be so cluttered. The usual rule of thumb is each attribute should be one value.

However, that's not really possible here. 

But, I'd certainly not include "Requester ID" on ever line. Probably only just the enabled requester IDs, maybe the index in hardware in a table

"%d\t%02x:%02x.%x"

I certainly wouldn't print the unenabled slots. That's not useful information and it'd be a pain to parse.

**Comment by kelvin-cao** [comment]
Updated per the comments. Thanks!
